src/sql/workbench/contrib/notebook/browser/notebook.contribution.ts

```diff
@@ -223,14 +223,16 @@ CommandsRegistry.registerCommand({
 
 		for (let editor of editors) {
 			if (editor instanceof NotebookInput) {
-				let model: INotebookModel = editor.notebookModel;
-				if (model?.providerId === JUPYTER_PROVIDER_ID) {
-					let jupyterNotebookManager: IExecuteManager = model.executeManagers.find(x => x.providerId === JUPYTER_PROVIDER_ID);
-					await jupyterNotebookManager.sessionManager.shutdownAll();
-					jupyterNotebookManager.sessionManager.dispose();
-					await jupyterNotebookManager.serverManager.stopServer();
-					return;
-				}
+				editor.modelResolved.then(async () => {
+					let model: INotebookModel = (<NotebookInput>editor).notebookModel;
+					if (model?.providerId === JUPYTER_PROVIDER_ID) {
+						let jupyterNotebookManager: IExecuteManager = model.executeManagers.find(x => x.providerId === JUPYTER_PROVIDER_ID);
+						await jupyterNotebookManager.sessionManager.shutdownAll();
+						jupyterNotebookManager.sessionManager.dispose();
+						await jupyterNotebookManager.serverManager.stopServer();
+						return;
```

Comment:
This is changing the logic, now it'll always loop through all editors regardless. To keep the previous behavior you could either wait for each editor to resolve (remove the .then and await, then keep the return)

Although looking at the original logic is this even correct? It would stop and return after the first editor it found with the Jupyter provider. Wouldn't you want to keep looping through all the other editors and shut down their servers as well?

Reply:
Originally I thought all python notebooks would be using the same Jupyter server but based on this comment [here](https://github.com/microsoft/azuredatastudio/blob/7804c87c870f738bb0ea452fdb5e6ecd68609294/extensions/notebook/src/jupyter/jupyterServerManager.ts#L110-L119) it is possible for there to be multiple Jupyter servers running so looping through all of the editors would be the correct behavior.

Reply:
Yeah good call out here. It's possible to have different Jupyter servers (e.g. notebooks opened from different drives).

Reply:
How is this going to handle multiple editors with the same server? If the first one is shut down will the others throw an error when they also try to get shut down?

Reply:
You're right, `sessionManager.shutdownAll()` is throwing an error because the server has already been shut down. I can't find a way to check if the server is closed from `jupyterSessionManager`. 

For now I've reverted the behavior back to before (stop and return after the first editor) since users having notebooks open from different drives is probably less common than having multiple notebooks open using the same jupyter server. 

I will keep looking for a way to shutdown multiple servers.

Reply:
I would also say it's reasonable to just log a warning but continue on if the server fails to shutdown. This seems like a best-effort kind of thing - if we can't shut down a server then we can probably still continue on (and maybe warn the user that something went wrong in case stuff later on doesn't work).

Do the sessions all share the same server object? Maybe we could have a status property on it if so - so when being told to shut down if it's already shutdown it just returns immediately.

Reply:
Tracking issue here https://github.com/microsoft/azuredatastudio/issues/17812